### PR TITLE
RFD 0167: Automatic updates change proposal

### PR DESCRIPTION
**Pull request by bernardjkim**
[Rendered](https://github.com/gravitational/teleport/blob/9820388363f8ec12a94123d1009db7ba570a7a76/rfd/0167-auto-updates-change-proposal.md)

This RFD proposes a change to the automatic updates design. The design has a number of limitations that are incompatible with the needs of Teleport Cloud. This RFD provides an overview of the current issues and some potential solutions. The RFD includes minimal implementation details. A separate execution plan with more more implementation details will be created, if the change proposals are approved.

Related Issues:
- https://github.com/gravitational/teleport/issues/39094
- https://github.com/gravitational/teleport/issues/39101
- https://github.com/gravitational/teleport/issues/39102
- https://github.com/gravitational/teleport/issues/39215

### CODE REVIEW

File: rfd/0167-auto-updates-change-proposal.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+Step 2 (Alternative): Instead of relying on the stable/vXX channels. There is also the option of preventing updates of the teleport-ent package except by the Teleport updater. This would not require the stable/cloud channel to be deprecated. However, this would require a different solution for each supported package manager, which could get complicated.
+
+Apt supports an apt-mark hold command that can be used to hold back a package from being updated. The Teleport updater can be modified to lock the teleport-ent package after an update, and unlock when it is performing an update.
+
+Yum has a similar feature that can exclude packages from a system update. This can be done by specifying teleport-ent to be excluded in the /etc/yum.conf file.
+
+Step 3: The Teleport installation process should no longer rely on the package manager to download teleport-ent packages. Instead, the Teleport proxy will now serve the latest compatible version of the teleport-ent package. The Teleport updater will then be responsible for downloading and installing the teleport-ent packages from the Teleport proxy.
```
Comment (fheinecke) [inline rfd/0167-auto-updates-change-proposal.md:135]: What's the plan for getting the artifacts securely delivered to the proxy?

File: rfd/0167-auto-updates-change-proposal.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+4. Teleport Cloud must provide the necessary tools for a user to build their own automation to keep Teleport client software up to date. This should include a stable API that users can rely on to request the latest version of Teleport that is compatible with their Teleport control plane.
+
+### Deprecate the stable/cloud teleport-ent package
+Currently, the teleport-ent-updater package requires the teleport-ent package as a dependency. This means that the user must install the latest version of the teleport-ent package which may or may not be compatible with their Teleport control plane, or they must first specify a compatible version of teleport-ent to install. This puts unnecessary burden on the user, and complicates the installation process.
+
+Step 1: To remove this burden from the user and simplify the installation process, the Teleport updater will support an install command. The install command accepts the necessary configuration and then installs the latest compatible version of the teleport-ent package for the user.
```
Comment (sclevine) [inline rfd/0167-auto-updates-change-proposal.md:115]: We discussed removing the package entirely. Are you still working on this section?

File: rfd/0167-auto-updates-change-proposal.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+![publishing](assets/0167-auto-updates-change-proposal/systemd-deployment.png)
+
+### Kubernetes Deployments
+Kubernetes deployments of Teleport are installed using Helm. The teleport/teleport-kube-agent Helm release deploys the teleport-agent resource and a separate teleport-agent-updater resource at the same version. The teleport-agent-updater periodically checks for available updates and updates the teleport-agent resource image version.
+
+![publishing](assets/0167-auto-updates-change-proposal/kube-deployment.png)
+
+## Current Limitations
+
```
Comment (hugoShaka) [inline rfd/0167-auto-updates-change-proposal.md:76]: I would like to add another current limitation: the automatic update process only covers the agents. Not the integrations.

Since v15 we have Teleport Operator users against Cloud instances, we already had users self-hosting plugins ([see slack link](https://goteleport.slack.com/archives/C0582MBSMHN/p1710343435263599?thread_ts=1710342970.418989&cid=C0582MBSMHN)).

I suspect that as Teleport Cloud grows and operator adoption increases, similar issues will arise. I think the proposed solution should be designed to be extended in the future to cover the integrations, at least self-hosted plugins and the operator.

File: rfd/0167-auto-updates-change-proposal.md (hunk 1 above)
Comment (hugoShaka) [reply rfd/0167-auto-updates-change-proposal.md:135]: I'm not sure we want to tunnel 500MiB blobs through the proxy to 10k agents in the same maintenance window. We'll have a bad availability, slow speed, pay a lot of bandwidth.

I would rather have the artifacts stored into an S3 bucket/CDN/whatever and the proxy to serve the address/version.

If we part from the package manager, we must solve the trust issue, exactly like we had to do for Kubernetes. The artifacts must be signed. The public key can be baked into the updater, or retrieved dynamically, maybe from the proxy. Serving the key from the proxy offers less guarantees if someone can MITM with valid TLS, but it would allow us to rotate the signing key without contacting all customers to redeploy.

Comment (sclevine) [reply rfd/0167-auto-updates-change-proposal.md:135]: > I would rather have the artifacts stored into an S3 bucket/CDN/whatever and the proxy to serve the bucket address and desired version.

Being able to curl off the proxy would be nice, but it should redirect to a CDN.

File: rfd/0167-auto-updates-change-proposal.md (hunk 3 above)
Comment (hugoShaka) [inline rfd/0167-auto-updates-change-proposal.md:76]: Another limitation Cloud is not facing directly is that automatic updates are cloud-only. Because AUs are not properly supported and adopted everywhere, in Teleport code we have a special execution path "if Cloud". This increases the difference between cloud and self-hosted, the probability of a bug happening. This path is typically less tested and expensive to maintain.

Getting rid of the special case would prevent incidents like the discovery one from 2 weeks ago.

File: rfd/0167-auto-updates-change-proposal.md (hunk 1 above)
Comment (hugoShaka) [reply rfd/0167-auto-updates-change-proposal.md:135]: I'm not sure we want to send the CDN URL from the proxy. This trades a lot of security guarantees against a bit of usability (if own control the domain, we can host whatever we want on it).

- if the agent reads the CDN URL from the proxy redirect: someone can takeover the proxy and have the agents install arbitrary things, especially if the proxy also serves the signature keys
- if the agent reads the version from the proxy but has a built-in CDN URL, someone taking over the proxy cannot force agents to install arbitrary content. Only content hosted in the teleport CDN can be downloaded. So it would require taking over the CDN as well (else the only thing you can do is a downgrade attack)

I think a middle ground would be having the proxy return the artifact name and the updater set the domain. This way, we can move artifacts and change the CDN structure remotely, but agents are not installing something random. It would also allow users to have the updater pull through a proxy/mirror/cache/whatever.

Comment (fheinecke) [reply rfd/0167-auto-updates-change-proposal.md:135]: What SLO do we have for auto updates and/or cloud? What SLO do we need to have for the release service (currently don't have any)?

Comment (sclevine) [reply rfd/0167-auto-updates-change-proposal.md:135]: > if the agent reads the CDN URL from the proxy redirect: someone can takeover the proxy and have the agents install arbitrary things, especially if the proxy also serves the signature keys

> So the installation method would be curl https://mytenant.teleport.sh/something/install.sh | bash

Difference here is just that it's less scary for initial install?

If I take over the proxy, I can serve arbitrary install and discover scripts.

Comment (hugoShaka) [reply rfd/0167-auto-updates-change-proposal.md:135]: > If I take over the proxy, I can serve arbitrary install and discover scripts.

Installation time and update time are not equivalent. If you take over a proxy, you can compromise new agents by changing the bash script served at install time. However, you cannot compromise existing agents, that's not a 0-click escalation.

This is already the case for regular joins, the agent just trusts whatever auth cert is returned by the proxy during theinitial join. The agent can be stolen if you have valid TLS certs. However, an already enrolled agent will check the cluster CAs and cannot be stolen. The automatic update mechanism should provide the same security guarantees.

Proving the bash script authenticity is hard because:
- we don't have established trust except the OS trust store
- the dynamic part injected by the proxy (version, URL, automatic updates, oss/ent, architecture, package manager, ...) blocks us from hashing the bash script

I'm not sure we have to solve the bootstrapping problem as Teleport doesn't even solve it for a regular agent join.

File: rfd/0167-auto-updates-change-proposal.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+### Self-Hosted Support
+The automatic updates feature has been pushed for Teleport Cloud clusters, but not for self-hosted clusters. This has lead development of special execution paths in the Teleport codebase just for Teleport Cloud. These paths are typically not test as extensively, and are expensive to maintain.
+
+Issue: This has lead to some incidents where changes made specifically for Teleport Cloud were merged but contained a bug in a specific edge case. The edge case was not thoroughly tested and the bug was deployed.
+
+### Integrations
+The automatic updates feature does not support Teleport integrations. The Teleport Operator was introduced in Teleport 15 and is already being used for self-hosted plugins. As adoption increases, users may face the same issues with keeping their integrations updated.
+
+## Change proposals
```
Comment (sclevine) [inline rfd/0167-auto-updates-change-proposal.md:111]: Things to add:
- Upgrader should not have its own timer. As soon as the proxy serves a new agent version, the agent is installed.
- The agent version served by the proxy should be configurable. I would suggest proxy read the version from a monitored file on disk, so that it's easy to update via Kubernetes / Cloud. tctl would work as well, but it's harder to change via Kubernetes.

File: rfd/0167-auto-updates-change-proposal.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+The Teleport updater must be able to resolve conflict for these two situations:
+1. If Teleport is already installed on the system via the system package manager, the Teleport updater must be able to migrate the Teleport service to use the new binary installed using the Teleport CDN.
+2. If Teleport is already installed into the /var/lib/teleport directory, a new installation of Teleport via the system package manager must not override the Teleport binary used by the Teleport service.
+
+Step 4: The Teleport documentation should be updated to include a new section with instructions about how a user can build their own update automation.
+
+### Agent Version Management
```
Comment (sclevine) [inline rfd/0167-auto-updates-change-proposal.md:141]: Can you add the proposed API for this?

Seems like we need at least three fields:
- Version
- Upgrade time OR immediate (can be a specific timestamp -- Cloud can translate window -> time)
- Jitter duration

And out-of-scope for this RFD, but eventually:
- Client bucket ID, for staged rollouts (upgrader holds off until the ID matches locally configured value)

File: rfd/0167-auto-updates-change-proposal.md [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+```sh
+$ apt-get install teleport-ent-updater
+$ teleport-upgrade install --proxy=example.teleport.sh
+```
+
+Step 2: Teleport must provide an alternate method of securely downloading the latest compatible version of Teleport. The Teleport CDN already serves Teleport artifacts and their SHA256 checksums. To ensure authenticity of these Teleport downloads, Teleport will now need to sign these artifacts and provide the digital signatures. The public key required to verify the digital signature will be baked into the Teleport updater.
+
+Step 3: The Teleport updater must control all Teleport updates. To ensure this, the Teleport updater will no longer rely on the system package manager to install/update Teleport. Users will no longer be able to manually update Teleport through the system package manager. Instead, the Teleport updater will now download the Teleport artifacts from the Teleport CDN along with the SHA256 checksum and the digital signature. The Teleport updater will verify the Teleport artifact and install Teleport into the /var/lib/teleport directory.
+
+The Teleport updater must be able to resolve conflict for these two situations:
```
[the remark this answers is not in the document]
Comment (bernardjkim) [reply rfd/0167-auto-updates-change-proposal.md:135]: > For existing installations, would it make sense to run teleport-upgrade install automatically in post-install scripts, given that we can detect the proxy address?

I was thinking the install script would run `teleport-upgrade install` right after installing the updater. Is what you're thinking a bit different? What would be the benefit of running `teleport-upgrade install` automatically in a post-install script?

> Should we remove the existing teleport package if it exists, or just ensure that the auto-upgrader version of teleport is executed by the systemd service?

Hmm, I'm thinking the second option would be more resilient? I'm thinking if the teleport package is still available for installation, there is probably a good chance that users reinstall the package unintentionally?

File: rfd/0167-auto-updates-change-proposal.md [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+Step 4: The Teleport documentation should be updated to include a new section with instructions about how a user can build their own update automation.
+
+### Agent Version Management
+Currently, the Teleport proxies rely on the upstream major version channels to serve the latest compatible version of Teleport to the updaters. This is configurable by modifying the teleport.yaml proxy configuration, but this requires pausing reconciliation for the resource because the Teleport Controller will revert the changes back to the default. Changing the proxy configuration also requires redeploying the proxy.
+
+The agent version should be made easier to configure using the Kubernetes or Cloud API. Modifying the agent version should not require reconciliation to be paused, and it should not require the Teleport proxy to be redeployed.
+
+A simple solution is to configure the Teleport proxies to now read the agent version from a monitored file on disk. Teleport Cloud will be able to easily and dynamically modify the agent version via the Kubernetes API.
```
[the remark this answers is not in the document]
Comment (bernardjkim) [reply rfd/0167-auto-updates-change-proposal.md:146]: Are we thinking we add configuration to this `autoupdate_version` resource proposed in the client tools rfd?
```
kind: autoupdate_version
spec:
  # tools_version is the version of client tools the cluster will advertise.
  # Can be auto (match the version of the proxy) or an exact semver formatted
  # version.
  tools_version: auto|X.Y.Z
```

**Comment by sclevine** [comment]
Superseded by https://github.com/gravitational/teleport/pull/39805.
Nearly all of the suggestions (including install script changes) were released in v17.3.

@bernardjkim closing this for now, but feel free to reopen if you have any concerns
